### PR TITLE
Broadcasting for semiclassical objects

### PR DESCRIPTION
**Pull request by apkille**
@david-pl @Krastanov @amilsted
- Added an interface for broadcasting `State` objects, which enables use of DiffEq solvers
- With these changes, ODE problems for `State`s can be naturally written as the following example:
```
using QuantumOptics, OrdinaryDiffEq, BenchmarkTools

b = SpinBasis(1//2)
psi0 = spindown(b)
u0 = ComplexF64[0.5, 0.75] 
sc = semiclassical.State(psi0, u0)
t₀, t₁ = (0.0, pi)
const σx = sigmax(b)

fquantum(t, q, u) = σx + cos(u[1])*identityoperator(σx)
fclassical!(du, u, q, t) = (du[1] = sin(u[2]); du[2] = 2*u[1])
f!(dstate, state, p, t) = semiclassical.dschroedinger_dynamic!(dstate, fquantum, fclassical!, state, t)
prob = ODEProblem(f!, sc, (t₀, t₁))

julia> @benchmark sol = solve($prob, DP5(); save_everystep=false)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  50.166 μs … 133.706 ms  ┊ GC (min … max):  0.00% … 99.93%
 Time  (median):     52.583 μs               ┊ GC (median):     0.00%
 Time  (mean ± σ):   70.621 μs ±   1.339 ms  ┊ GC (mean ± σ):  23.35% ±  3.73%

       ▃█▅                                                      
  ▁▂▃▆████▇▃▂▂▂▂▂▂▂▂▂▃▃▃▃▂▃▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  50.2 μs         Histogram: frequency by time         69.8 μs <

 Memory estimate: 93.81 KiB, allocs estimate: 1136.

julia> @benchmark semiclassical.schroedinger_dynamic([$t₀, $t₁], $sc, $fquantum, $fclassical!)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  117.875 μs … 137.310 ms  ┊ GC (min … max):  0.00% … 99.87%
 Time  (median):     122.959 μs               ┊ GC (median):     0.00%
 Time  (mean ± σ):   148.349 μs ±   1.379 ms  ┊ GC (mean ± σ):  15.47% ±  6.25%

    ▃█▇▁▁▃▃                                                      
  ▂▅████████▄▃▃▂▂▃▃▂▃▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  118 μs           Histogram: frequency by time          163 μs <

 Memory estimate: 225.66 KiB, allocs estimate: 2637.
```
- Support is provided for both kets and density operators.
- *Note*: broadcasting for a custom type like `State` is weird, and really the only reason we're doing it is to have interoperability of semiclassical objects with SciML. 
- Some of these interface decisions probably can be debated. I based most of them off of how `State` objects were recasted in the `semiclassical.schroedinger_dynamic`, `semiclassical.master_dynamic`, etc. functions.

### CODE REVIEW

File: src/semiclassical.jl
Comment (Krastanov) [inline src/semiclassical.jl:131]: Same comment about this import being brought up to the root file and turned into an `import` as in QuantumOpticsBase.jl#172

**Review by Krastanov — CHANGES_REQUESTED** [review]
Looks good to merge after a small import change.

**Comment by Krastanov** [comment]
And after the errors in the tests are fixed. Does this depend on the changes to QuantumOpticsBase in its broadcast PR? If so, we should bump the compat bound for it.

**Comment by Krastanov** [comment]
And same comment as in the QuantumOpticsBase PR on adding your examples from the PR description to the test runner, maybe as a new file.

**Comment by apkille** [comment]
> And after the errors in the tests are fixed. Does this depend on the changes to QuantumOpticsBase in its broadcast PR? If so, we should bump the compat bound for it.

Yes, so we should merge this after we merge the QuantumOpticsBase PR.

**Comment by apkille** [comment]
@Krastanov Should be good to go.

**Comment by apkille** [comment]
@krastanov is there anything preventing this from getting merged?

File: src/semiclassical.jl
Comment (Krastanov) [inline src/semiclassical.jl:61]: the `kwargs` are lost

**Review by Krastanov — CHANGES_REQUESTED** [review]
This looks good to me.

There is a minor issue with `isapprox` not propagating its kwargs.

More importantly though, there are a few functions that codecov reports as not covered by tests. Could you add new tests for them, just to ensure the review is not missing anything? I am a bit skittish here as this touches a lot of basic interfaces.

**Comment by apkille** [comment]
@krastanov good catch. Should be covered now.

**Pull request merged**
